### ISSUE TITLE
98% cpu usage due to recursion

### ISSUE THREAD

**Issue by hippalus** [ahmetaa/zemberek-nlp#263]
The **dfs** **recursive method** in the **WordSegmenter** class causes high CPU usage in some scenarios.
See below screenshot of JVM profiler and CPU profiler.

![2021-03-16  21 06 18](https://user-images.githubusercontent.com/17534654/111358563-85562400-869b-11eb-9ede-54fefb97d8d2.jpg)

![image](https://user-images.githubusercontent.com/17534654/111358151-137dda80-869b-11eb-99b3-21e21488fadd.png)

**Comment by ahmetaa** [comment]
Yes, WordSegmenter uses a bad algorithm that works ok in most cases but fails miserably in some cases. It should have been using a dynamic programming approach but sadly it is not. Unfortunately I do not have much spare time to fix this, I would advise using a different algorithm for this.

**Comment by ahmetaa** [comment]
One idea that may alleviate this issue is to split input from spaces before processing with this. Also please provide an example that may cause this bad recursion.
